app/data_models/supplementary_data_store.py

```diff
@@ -0,0 +1,121 @@
+from __future__ import annotations
+
+from functools import cached_property
+from typing import Iterable, Mapping, MutableMapping, TypeAlias
+
+from werkzeug.datastructures import ImmutableDict
+
+SupplementaryDataKeyType: TypeAlias = tuple[str, str | None]
+SupplementaryDataValueType: TypeAlias = dict | str | list | None
+
+
+class InvalidSupplementaryDataSelector(Exception):
+    pass
+
+
+class SupplementaryDataStore:
+    """
+    An object that stores supplementary data
+    """
+
+    def __init__(
+        self,
+        supplementary_data: MutableMapping | None = None,
+        list_mappings: Mapping[str, Mapping] | None = None,
+    ):
+        """
+        Initialised with the "data" value from the sds api response
+        and list mappings of the form { list_name: { identifier : list_item_id }}
+        """
+        self._raw_data = supplementary_data or {}
+        self._list_mappings = list_mappings or {}
+        # use shallow copy of the data, as items will be popped off
+        self._data_map = self._build_map({**self._raw_data})
+
+    @cached_property
+    def raw_data(self) -> ImmutableDict:
+        return ImmutableDict(self._raw_data)
```

Comment:
It would still be mutable right, as this wouldn't make the nested dicts an ImmutableDict?. We have a `make_immutable` helper.

Reply:
Yep, good point, have switched to using `make_immutable` instead